### PR TITLE
Report generator: CSV parsing and HTML frame

### PR DESCRIPTION
**Pull request by dnoursi**
This is a WIP for how I propose to do report generation. It is missing some pieces, but fit for broad critique/input as I continue working on it.

The first step is complete here: putting CSVs for each trial into a single CSV for a given `benchmark/fengine`. 

Some work which remains is to put all fengine CSVs into a single CSV for each `benchmark/` folder. This could be the maximum trial from within each fengine CSV, or the median trial, or just all of them.

There also remains copying `setOfCharts.html` into the appropriate directories, as well as the creation of a `main.html` which would simply `iframe` all respective copies of `setOfCharts.html`. I believe `main.html` could be generated in golang, as it would be a series of very similar/formulaic lines of code.

### CODE REVIEW

File: engine-comparison/report-gen/generate-report.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		for k := 1; k < recordCols; k++ {
+			records[j][k] = ""
+		}
+	}
+	return records
+}
+
+// Enters all report subdirectories, from benchmark to fengine to trial;
+// composes individual CSVs (only two columns) into larger CSVs
+func composeAllNamed(desired_report_fname string) {
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: This is a long function with deep nesting.  Please use function calls within the for loops to reduce nesting and shorten the code.

File: engine-comparison/report-gen/generate-report.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	"path"
+        "io/ioutil"
+//        "strings"
+        "strconv"
+        "encoding/csv"
+        "os"
+)
+func checkErr(e error){
+	if (e != nil) {
+		fmt.Println("POOP!")
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: A more descriptive error message would be nice.

File: engine-comparison/report-gen/generate-report.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		os.Exit(1)
+	}
+}
+
+// Removes non-directory elements of any []os.FileInfo (a helper function, for convenience)
+// Makes use of idiomatic Golang return: out_ls is declared and returned implicitly
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Unnecessary comment.  Standard part of Go.

File: engine-comparison/report-gen/generate-report.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        "os"
+)
+func checkErr(e error){
+	if (e != nil) {
+		fmt.Println("POOP!")
+		os.Exit(1)
+	}
+}
+
+// Removes non-directory elements of any []os.FileInfo (a helper function, for convenience)
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: > (a helper function, for convenience)

Unnecessary info.  This is clear from the context.

File: engine-comparison/report-gen/generate-report.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+				checkErr(err)
+				// Add the name of this new column to records[0]
+				records[0] = append(records[0], fengine.Name() + trial.Name())
+
+				for _, row := range experiment_records {
+					// row[0] is time, on the x-axis; row[1] is value, on the y-axis
+					time_now, err := strconv.Atoi(row[0])
+					checkErr(err)
+					//If this test went longer than all of the others, so far
+					if(len(records) < time_now + 1) {
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: We should do this check once per trial rather than within the loop.  Just look at the last time in experiment_records and extend records accordingly.  Then start the loop.

File: engine-comparison/report-gen/generate-report.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+// Extend "records" matrix to have rows until time "desired_time"
+// Return: Extended version of record
+func extendRecordsToTime(records [][]string, desired_time int, recordCols int) ([][]string)  {
+	lenr := len(records)
+	// records[1] stores cycle [1], as records[0] is column names
+	for j := lenr; j < desired_time + 1; j++ {
+		records = append(records, make([]string, recordCols))
+		records[j][0] = strconv.Itoa(j)
+		for k := 1; k < recordCols; k++ {
+			records[j][k] = ""
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Redundant.  I believe make already sets default empty strings.

**Review by morehouse — COMMENTED** [review]
General comment:  Please format your Go files with gofmt.

File: engine-comparison/report-gen/generate-report.go (hunk 6 above)
Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: I had a couple issues with `append` because strings are already lists, but you're definitely right about this, thanks

File: engine-comparison/report-gen/generate-report.go (hunk 2 above)
Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: I agree, in practice this function hasn't come up much. Printing `e` is reasonable

File: engine-comparison/report-gen/generate-report.go (hunk 1 above)
Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: In general, 4 nested loops is a lot, and external functions are preferable. But in working on this, I've encountered that the inner sections of code rely kind of heavily on previous code, so externalizing this code requires passing sometimes 5-6 arguments to external functions

**Comment by dnoursi** [comment]
I've just abstracted the innermost loop to it's own function, but the other 3 don't have very much going on inside.

**Comment by morehouse** [comment]
```
composeAllNamed(...) {
  for bmark in bmarks {
    path = join(master_path, bmark.Name())
    handleBmark(desired_report_fname, path)
  }
}

handleBmark(...) {
  ...
  for fengine in fengines {
    path = join(path, fengine.Name())
    handleFengine(desired_report_fname, path)
  }
}

handleFengine(...) {
  ...
  records = ...
  num_record_columns = ...
  for j, trial in trials {
    read_report_path = join(path, trial.Name(), desired_report_fname)
    handleTrial(read_report_path, records, j, num_record_columns)
  }
  this_fe_writer = ...
  this_fe.writer.WriteAll(records)
}
```
Seems fairly easy to abstract more without getting long parameter lists.

**Comment by dnoursi** [comment]
This worked pretty nicely, and in the end most arguments were only for the final step of handling trials

File: engine-comparison/report-gen/generate-report.go [hunk 7]
```diff
@@ -66,56 +63,63 @@ func handleTrialCSV(this_reader *csv.Reader, records [][]string, fengine os.File
...
+
+func handleBmark(bmark os.FileInfo, current_path string, desired_report_fname string) {
+	potential_fengines, err := ioutil.ReadDir(path.Join(current_path, bmark.Name()))
+	checkErr(err)
+	// narrow potential_fengines to fengines so the indices of `range fengines` are useful
+	fengines := onlyDirectories(potential_fengines)
+
+	for _, fengine := range fengines {
+		handleFengine(fengine, path.Join(current_path, bmark.Name()), desired_report_fname)
+
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Unnecessary whitespace

File: engine-comparison/report-gen/generate-report.go [hunk 8]
```diff
@@ -66,56 +63,63 @@ func handleTrialCSV(this_reader *csv.Reader, records [][]string, fengine os.File
...
+	num_record_columns := len(trials) + 1
+	for j, trial := range trials {
+		// Create fds
+		this_file, err := os.Open(path.Join(current_path, fengine.Name(), trial.Name(), desired_report_fname))
+		checkErr(err)
+		defer this_file.Close()
+		this_reader := csv.NewReader(this_file)
+
+		records = handleTrialCSV(this_reader, records, fengine, trial, num_record_columns, j)
+
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Unnecessary whitespace

File: engine-comparison/report-gen/generate-report.go [hunk 9]
```diff
@@ -66,56 +63,63 @@ func handleTrialCSV(this_reader *csv.Reader, records [][]string, fengine os.File
 	return records
 }
 
+func handleFengine(fengine os.FileInfo, current_path string, desired_report_fname string) {
+	this_fe_file, err := os.Create(path.Join(current_path, fengine.Name(), desired_report_fname))
+	checkErr(err)
+	defer this_fe_file.Close()
+	this_fe_writer := csv.NewWriter(this_fe_file)
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Open this file and writer closer to where it is used.

File: engine-comparison/report-gen/generate-report.go [hunk 10]
```diff
@@ -66,56 +63,63 @@ func handleTrialCSV(this_reader *csv.Reader, records [][]string, fengine os.File
...
+	potential_trials, err := ioutil.ReadDir(path.Join(current_path, fengine.Name()))
+	checkErr(err)
+	trials := onlyDirectories(potential_trials)
+
+	num_record_columns := len(trials) + 1
+	for j, trial := range trials {
+		// Create fds
+		this_file, err := os.Open(path.Join(current_path, fengine.Name(), trial.Name(), desired_report_fname))
+		checkErr(err)
+		defer this_file.Close()
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Depending on the number of trials here, we could be opening a lot of files that remain open until handleFengine() returns.

File: engine-comparison/report-gen/generate-report.go [hunk 11]
```diff
@@ -66,56 +63,63 @@ func handleTrialCSV(this_reader *csv.Reader, records [][]string, fengine os.File
...
+		this_reader := csv.NewReader(this_file)
+
+		records = handleTrialCSV(this_reader, records, fengine, trial, num_record_columns, j)
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: fengine and trial can be combined into a single column_header parameter.

File: engine-comparison/report-gen/generate-report.go (hunk 10 above)
Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: Certainly true; `defer` seems to be strictly less efficient, and is primarily useful for readability I suppose. Definitely worse in a loop

File: engine-comparison/report-gen/generate-report.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+		time_now, err := strconv.Atoi(row[0])
+		checkErr(err)
+		records[time_now][j+1] = row[1]
+	}
+	return records
+}
+
+func handleFengine(fengine os.FileInfo, current_path string, desired_report_fname string) {
+
+
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Please remove unnecessary whitespace at beginning or end of functions.

**Comment by dnoursi** [comment]
That should be all for Golang code, for now - `appendFastestTrial` could always be complemented with other functions, such as `appendMedianTrial` or `appendAllTrials`. For now, though, hopefully this PR is done!

File: engine-comparison/report-gen/generate-report.go [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+		meta_records = extendRecordsToTime(meta_records, speed_of_ft, len(meta_records[0]))
+	}
+	for j := 0; j <= speed_of_ft; j++ {
+		meta_records[j] = append(meta_records[j], records[j][fastest_trial])
+	}
+	return meta_records
+}
+
+// Call handleFengine() for each fengine, then compose all fengine data into a single CSV for comparison
+func handleBmark(bmark os.FileInfo, current_path string, desired_report_fname string) {
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Every time you use bmark here, you append it to current_path.  So you should just pass the path to the benchmark instead.

Also, bmark_path is more descriptive here.

File: engine-comparison/report-gen/generate-report.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+	for _, fengine := range fengines {
+		records := handleFengine(fengine, path.Join(current_path, bmark.Name()), desired_report_fname)
+		meta_records = appendFastestTrial(meta_records, records)
+	}
+	this_bm_file, err := os.Create(path.Join(current_path, bmark.Name(), desired_report_fname))
+	checkErr(err)
+	this_bm_writer := csv.NewWriter(this_bm_file)
+	this_bm_writer.WriteAll(meta_records)
+	this_bm_file.Close()
+
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Remove unnecessary whitespace from beginning and end of functions.

File: engine-comparison/report-gen/generate-report.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+	for _, row := range experiment_records {
+		// row[0] is time, on the x-axis; row[1] is value, on the y-axis
+		time_now, err := strconv.Atoi(row[0])
+		checkErr(err)
+		records[time_now][j+1] = row[1]
+	}
+	return records
+}
+
+func handleFengine(fengine os.FileInfo, current_path string, desired_report_fname string) [][]string {
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: current_path isn't very descriptive. bmark_path seems more intuitive.

File: engine-comparison/report-gen/generate-report.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+	// records[1] stores cycle [1], as records[0] is column names
+	for j := lenr; j < desired_time+1; j++ {
+		records = append(records, make([]string, record_cols))
+		records[j][0] = strconv.Itoa(j)
+	}
+	return records
+}
+
+// Handles the CSV Reader for a single trial, and updates records[][] accordingly. Returns the updated records
+func handleTrialCSV(this_reader *csv.Reader, records [][]string, column_name string, num_record_columns int, j int) [][]string {
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Most parameter names aren't very descriptive here.  Some suggested improvements:
- this_reader -> trial_reader
- records -> fengine_records
- num_record_columns -> num_total_cols
- j -> col_index

File: engine-comparison/report-gen/generate-report.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+func handleBmark(bmark os.FileInfo, current_path string, desired_report_fname string) {
+	meta_records := [][]string{{"time"}}
+
+	potential_fengines, err := ioutil.ReadDir(path.Join(current_path, bmark.Name()))
+	checkErr(err)
+	// narrow potential_fengines to fengines so the indices of `range fengines` are useful
+	fengines := onlyDirectories(potential_fengines)
+
+	for _, fengine := range fengines {
+		records := handleFengine(fengine, path.Join(current_path, bmark.Name()), desired_report_fname)
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: I think fengine_records is more descriptive.

File: engine-comparison/report-gen/generate-report.go [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+	}
+	for j := 0; j <= speed_of_ft; j++ {
+		meta_records[j] = append(meta_records[j], records[j][fastest_trial])
+	}
+	return meta_records
+}
+
+// Call handleFengine() for each fengine, then compose all fengine data into a single CSV for comparison
+func handleBmark(bmark os.FileInfo, current_path string, desired_report_fname string) {
+	meta_records := [][]string{{"time"}}
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: I think bmark_records is more descriptive.

File: engine-comparison/report-gen/generate-report.go [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+func appendFastestTrial(meta_records [][]string, records [][]string) [][]string {
+	// Initialized to false: track whether this trial has lasted a long time
+	trials := make([]bool, len(records[0])-1)
+	// Use int for faster comparisons
+	finished_trials := len(trials)
+	// Speed of Fastest Trial
+	var speed_of_ft int
+	var fastest_trial int
+
+	// Find the fastest trial by working backwards, since data points are sometimes dropped
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: When are data points dropped?  What if data points at the end are dropped?

File: engine-comparison/report-gen/generate-report.go [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+	for r := len(records) - 1; r > 0; r-- {
+		for c := 1; c < len(records[r]); c++ {
+			if (trials[c-1] == false) && (len(records[r][c]) > 0) {
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: `!trials[c-1] && len(records[r][c]) > 0`

File: engine-comparison/report-gen/generate-report.go [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// Identify the fastest trial  column in records and add it to meta_records
+func appendFastestTrial(meta_records [][]string, records [][]string) [][]string {
+	// Initialized to false: track whether this trial has lasted a long time
+	trials := make([]bool, len(records[0])-1)
+	// Use int for faster comparisons
+	finished_trials := len(trials)
+	// Speed of Fastest Trial
+	var speed_of_ft int
```
Comment (morehouse) [inline engine-comparison/report-gen/generate-report.go]: Speed isn't a very precise description.  `rows_in_fastest_trial` seems more appropriate.

File: engine-comparison/report-gen/generate-report.go (hunk 19 above)
Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: Data is dropped in the runner if `gsutil` takes more time than one cycle to sync the corpus. Data points at the end shouldn't be dropped, because that's the last `rsync` before resetting for another trial. However, looking at `runner.sh` now, I think this last cycle might accidentally be dropped every time

File: engine-comparison/report-gen/generate-report.go (hunk 13 above)
Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: I kind of prefer passing `bmark`,  to keep the options open for further record creation, as well as (less importantly) to mirror `handleFengine` and

Comment (morehouse) [reply engine-comparison/report-gen/generate-report.go]: In that case, join the paths once at the top of the function rather than 3 different times.

File: engine-comparison/report-gen/generate-report.go (hunk 19 above)
Comment (morehouse) [reply engine-comparison/report-gen/generate-report.go]: Is the issue that the second `gsutil rsync` overrides the first or that the dispatcher doesn't check the same cycle again if `gsutil` is still going?

Comment (dnoursi) [reply engine-comparison/report-gen/generate-report.go]: Sort of the first; `gsutil` doesn't run in the background on runners, so the while loop doesn't continue to snapshot the next corpus until `gsutil` finishes. If it did run in the background, there would be the possibility of two `gsutil rsync`s running at the same time, trying to sync two different files to the same directory. Which could perhaps work, but would likely just further overload internet bandwidth

Comment (morehouse) [reply engine-comparison/report-gen/generate-report.go]: Ok, I see.

**Review by morehouse — APPROVED** [review]
